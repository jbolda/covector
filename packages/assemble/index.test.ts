import { assemble, mergeIntoConfig, mergeChangesToConfig } from "./index";
import fixtures from "fixturez";
const f = fixtures(__dirname);

const vfilePart = (filename: string) => ({
  path: "",
  extname: "",
  data: { filename },
});

const assembledChanges = {
  releases: {
    "@namespaced/assemble2": {
      changes: [
        {
          releases: {
            "@namespaced/assemble2": "patch",
            assemble1: "patch",
          },
          summary: "This is a test.",
        },
      ],
      type: "patch",
    },
    assemble1: {
      changes: [
        {
          releases: {
            assemble1: "patch",
            assemble2: "patch",
          },
          summary: "This is a test.",
        },
        {
          releases: {
            assemble1: "minor",
            assemble2: "patch",
          },
          summary: "This is a test.",
        },
        {
          releases: {
            assemble1: "patch",
            assemble2: "major",
          },
          summary: "This is a test.",
        },
        {
          releases: {
            "@namespaced/assemble2": "patch",
            assemble1: "patch",
          },
          summary: "This is a test.",
        },
      ],
      type: "minor",
    },
    assemble2: {
      changes: [
        {
          releases: {
            assemble1: "patch",
            assemble2: "patch",
          },
          summary: "This is a test.",
        },
        {
          releases: {
            assemble1: "minor",
            assemble2: "patch",
          },
          summary: "This is a test.",
        },
        {
          releases: {
            assemble1: "patch",
            assemble2: "major",
          },
          summary: "This is a test.",
        },
      ],
      type: "major",
    },
  },
};

const config = {
  pkgManagers: {
    javascript: {
      version: "lerna version ${ release.type } --no-git-tag-version --no-push",
      publish: "npm publish",
    },
  },
  packages: {
    assemble1: {
      path: "./packages/assemble1",
      manager: "javascript",
      dependencies: ["all"],
    },
    assemble2: {
      path: "./packages/assemble2",
      version: "lerna version ${ release.type }",
      dependencies: ["all"],
    },
    "@namespaced/assemble1": {
      path: "./packages/namespaced-assemble2",
      manager: "cargo",
      version: "cargo version ${ release.type }",
      publish: "cargo publish",
      dependencies: ["assemble1", "all"],
    },
    "@namespaced/assemble2": {
      path: "./packages/namespaced-assemble2",
      manager: "cargo",
      version: "cargo version ${ release.type }",
      publish: "cargo publish",
      dependencies: ["assemble2", "all"],
    },
    all: {
      version: true,
    },
  },
};

const configSpecial = {
  ...config,
  additionalBumpTypes: ["housekeeping", "workflows"],
};

const testTextOne = {
  ...vfilePart("testTextOne"),
  contents: `
---
"assemble1": patch
"assemble2": patch
---
    
This is a test.
`,
};
const testTextTwo = {
  ...vfilePart("testTextTwo"),
  contents: `
---
"assemble1": minor
"assemble2": patch
---
    
This is a test.
`,
};
const testTextThree = {
  ...vfilePart("testTextThree"),
  contents: `
---
"assemble1": patch
"assemble2": major
---
    
This is a test.
`,
};
const testTextFour = {
  ...vfilePart("testTextFour"),
  contents: `
---
"assemble1": patch
"@namespaced/assemble2": patch
---
    
This is a test. We might link out to a [website](https://www.jacobbolda.com).
`,
};
const testTextFive = {
  ...vfilePart("testTextFive"),
  contents: `
---
"all": minor
---
  
This is a test.
`,
};
const testTextSpecialOne = {
  ...vfilePart("testTextSpecialOne"),
  contents: `
---
"assemble1": housekeeping
"assemble2": workflows
---
  
This is a test.
`,
};
const testTextSpecialTwo = {
  ...vfilePart("testTextSpecialTwo"),
  contents: `
---
"assemble1": patch
"assemble2": workflows
"@namespaced/assemble2": explosions
---
  
This is a test.
`,
};

describe("assemble changes", () => {
  it("runs", function* (): Generator<any> {
    const assembled = yield assemble({
      vfiles: [testTextOne, testTextTwo, testTextThree, testTextFour],
    });
    expect(assembled).toMatchSnapshot();
  });

  it("assembles deps", function* (): Generator<any> {
    const assembled = yield assemble({ vfiles: [testTextFive] });
    expect(assembled).toMatchSnapshot();
  });
});

describe("assemble changes in preMode", () => {
  it("with no existing changes", function* (): Generator<any> {
    const assembled = yield assemble({
      vfiles: [testTextOne, testTextTwo, testTextThree, testTextFour],
      preMode: { on: true, prevFiles: [] },
    });
    expect(assembled).toMatchSnapshot();
  });

  it("with existing changes that upgrade", function* (): Generator<any> {
    const assembled = yield assemble({
      vfiles: [testTextOne, testTextTwo, testTextThree, testTextFour],
      preMode: { on: true, prevFiles: [testTextOne.data.filename] },
    });
    expect(assembled).toMatchSnapshot();
  });

  it("with existing changes with the same bump", function* (): Generator<any> {
    const assembled = yield assemble({
      vfiles: [testTextOne, testTextTwo, testTextFour],
      preMode: { on: true, prevFiles: [testTextOne.data.filename] },
    });
    expect(assembled).toMatchSnapshot();
  });
});

describe("errors on bad change files", () => {
  const emptyChangefile = {
    path: "",
    extname: "",
    data: { filename: ".changes/empty-file.md" },
    contents: `---
---
  
This doesn't bump much.
`,
  };

  it("throws on no changes", function* (): Generator<any> {
    expect.assertions(1);
    try {
      yield assemble({
        vfiles: [emptyChangefile],
      });
    } catch (e) {
      expect(e.message).toMatch(
        ".changes/empty-file.md didn't have any packages bumped. Please add a package bump."
      );
    }
  });
});

describe("special bump types", () => {
  it("valid additional bump types", function* (): Generator<any> {
    const assembled = yield assemble({
      vfiles: [
        testTextOne,
        testTextTwo,
        testTextThree,
        testTextFour,
        testTextSpecialOne,
      ],
      //@ts-ignore
      config: configSpecial,
    });
    expect(assembled).toMatchSnapshot();
  });

  it("invalid bump types", function* (): Generator<any> {
    expect.assertions(1);
    try {
      yield assemble({
        vfiles: [
          testTextOne,
          testTextTwo,
          testTextThree,
          testTextFour,
          testTextSpecialOne,
        ],
        //@ts-ignore
        config,
      });
    } catch (e) {
      expect(e.message).toMatch(
        "housekeeping specified for assemble1 is invalid.\n" +
          "Try one of the following: major, minor, patch.\n"
      );
    }
  });

  it("one each valid and invalid", function* (): Generator<any> {
    expect.assertions(1);
    try {
      yield assemble({
        vfiles: [testTextSpecialTwo],
        //@ts-ignore
        config: configSpecial,
      });
    } catch (e) {
      expect(e.message).toMatch(
        "explosions specified for @namespaced/assemble2 is invalid.\n" +
          "Try one of the following: major, minor, patch, housekeeping, workflows.\n"
      );
    }
  });

  it("handles an only noop", function* (): Generator<any> {
    const assembled = yield assemble({
      vfiles: [testTextSpecialOne],
      //@ts-ignore
      config: configSpecial,
    });
    expect(assembled).toMatchSnapshot();
  });
});

describe("merge config test", () => {
  it("merges version", function* (): Generator<any> {
    const mergedVersionConfig = yield mergeChangesToConfig({
      //@ts-ignore
      config,
      assembledChanges,
      command: "version",
    });
    expect(mergedVersionConfig).toMatchSnapshot();
  });

  it("merges version without command", function* (): Generator<any> {
    let modifiedConfig = { ...config };
    //@ts-ignore
    delete modifiedConfig.pkgManagers.javascript.version;
    //@ts-ignore
    delete modifiedConfig.packages["assemble2"].version;
    //@ts-ignore
    delete modifiedConfig.packages["@namespaced/assemble1"].version;
    //@ts-ignore
    delete modifiedConfig.packages["@namespaced/assemble2"].version;

    const mergedVersionConfig = yield mergeChangesToConfig({
      //@ts-ignore
      config: modifiedConfig,
      assembledChanges,
      command: "version",
    });
    expect(mergedVersionConfig).toMatchSnapshot();
  });

  it("merges nested bumps", function* (): Generator<any> {
    const nestedAssembledChanges = {
      releases: {
        assemble1: {
          changes: [
            {
              releases: {
                assemble1: "patch",
              },
              summary: "This is a test.",
            },
          ],
          type: "minor",
        },
        assemble2: {
          changes: [
            {
              releases: {
                all: "minor",
              },
              summary: "This is a test.",
            },
          ],
          type: "minor",
        },
        all: {
          changes: [
            {
              releases: {
                all: "minor",
              },
              summary: "This is a test.",
            },
          ],
          type: "minor",
        },
      },
    };

    const nestedConfig = {
      packages: {
        assemble1: {
          path: "./packages/assemble1",
          version: true,
          dependencies: ["assemble1", "all"],
        },
        assemble2: {
          path: "./packages/assemble2",
          version: true,
          dependencies: ["all"],
        },
        all: {
          version: true,
        },
      },
    };

    const mergedVersionConfig = yield mergeChangesToConfig({
      //@ts-ignore
      config: nestedConfig,
      assembledChanges: nestedAssembledChanges,
      command: "version",
    });
    expect(mergedVersionConfig).toMatchSnapshot();
  });

  it("merges publish", function* (): Generator<any> {
    const configFolder = f.copy("assemble");

    const mergedPublishConfig = yield mergeIntoConfig({
      cwd: configFolder,
      //@ts-ignore
      config,
      //@ts-ignore
      assembledChanges: [],
      command: "publish",
    });
    expect(scrubVfile(mergedPublishConfig)).toMatchSnapshot();
  });
});

describe("merge filtered config test", () => {
  it("merges version", function* (): Generator<any> {
    const mergedVersionConfig = yield mergeChangesToConfig({
      //@ts-ignore
      config,
      assembledChanges,
      command: "version",
      filterPackages: ["assemble1", "@namespaced/assemble1"],
    });
    expect(mergedVersionConfig).toMatchSnapshot();
  });

  it("merges publish", function* (): Generator<any> {
    const configFolder = f.copy("assemble");

    const mergedPublishConfig = yield mergeIntoConfig({
      cwd: configFolder,
      //@ts-ignore
      config,
      assembledChanges,
      command: "publish",
      filterPackages: ["assemble1", "@namespaced/assemble1"],
    });
    expect(scrubVfile(mergedPublishConfig)).toMatchSnapshot();
  });
});

// vfile returns fs information that is flaky between machines, scrub it
const scrubVfile = (mergedPublishConfig: any) => {
  return mergedPublishConfig.commands.map((pkg: any) => {
    delete pkg.pkgFile.vfile;
    return pkg;
  }, mergedPublishConfig.commands);
};
