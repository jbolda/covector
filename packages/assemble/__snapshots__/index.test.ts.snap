// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`assemble changes assembles deps 1`] = `
Object {
  "changes": Array [
    Object {
      "releases": Object {
        "all": "minor",
      },
      "summary": "This is a test.",
    },
  ],
  "releases": Object {
    "all": Object {
      "changes": Array [
        Object {
          "releases": Object {
            "all": "minor",
          },
          "summary": "This is a test.",
        },
      ],
      "type": "minor",
    },
  },
}
`;

exports[`assemble changes in preMode with existing changes that upgrade 1`] = `
Object {
  "changes": Array [
    Object {
      "releases": Object {
        "assemble1": "minor",
        "assemble2": "patch",
      },
      "summary": "This is a test.",
    },
    Object {
      "releases": Object {
        "assemble1": "patch",
        "assemble2": "major",
      },
      "summary": "This is a test.",
    },
    Object {
      "releases": Object {
        "@namespaced/assemble2": "patch",
        "assemble1": "patch",
      },
      "summary": "This is a test. We might link out to a [website](https://www.jacobbolda.com).",
    },
  ],
  "releases": Object {
    "@namespaced/assemble2": Object {
      "changes": Array [
        Object {
          "releases": Object {
            "@namespaced/assemble2": "patch",
            "assemble1": "patch",
          },
          "summary": "This is a test. We might link out to a [website](https://www.jacobbolda.com).",
        },
      ],
      "type": "prepatch",
    },
    "assemble1": Object {
      "changes": Array [
        Object {
          "releases": Object {
            "assemble1": "minor",
            "assemble2": "patch",
          },
          "summary": "This is a test.",
        },
        Object {
          "releases": Object {
            "assemble1": "patch",
            "assemble2": "major",
          },
          "summary": "This is a test.",
        },
        Object {
          "releases": Object {
            "@namespaced/assemble2": "patch",
            "assemble1": "patch",
          },
          "summary": "This is a test. We might link out to a [website](https://www.jacobbolda.com).",
        },
      ],
      "type": "preminor",
    },
    "assemble2": Object {
      "changes": Array [
        Object {
          "releases": Object {
            "assemble1": "minor",
            "assemble2": "patch",
          },
          "summary": "This is a test.",
        },
        Object {
          "releases": Object {
            "assemble1": "patch",
            "assemble2": "major",
          },
          "summary": "This is a test.",
        },
      ],
      "type": "premajor",
    },
  },
}
`;

exports[`assemble changes in preMode with existing changes with the same bump 1`] = `
Object {
  "changes": Array [
    Object {
      "releases": Object {
        "assemble1": "minor",
        "assemble2": "patch",
      },
      "summary": "This is a test.",
    },
    Object {
      "releases": Object {
        "@namespaced/assemble2": "patch",
        "assemble1": "patch",
      },
      "summary": "This is a test. We might link out to a [website](https://www.jacobbolda.com).",
    },
  ],
  "releases": Object {
    "@namespaced/assemble2": Object {
      "changes": Array [
        Object {
          "releases": Object {
            "@namespaced/assemble2": "patch",
            "assemble1": "patch",
          },
          "summary": "This is a test. We might link out to a [website](https://www.jacobbolda.com).",
        },
      ],
      "type": "prepatch",
    },
    "assemble1": Object {
      "changes": Array [
        Object {
          "releases": Object {
            "assemble1": "minor",
            "assemble2": "patch",
          },
          "summary": "This is a test.",
        },
        Object {
          "releases": Object {
            "@namespaced/assemble2": "patch",
            "assemble1": "patch",
          },
          "summary": "This is a test. We might link out to a [website](https://www.jacobbolda.com).",
        },
      ],
      "type": "preminor",
    },
    "assemble2": Object {
      "changes": Array [
        Object {
          "releases": Object {
            "assemble1": "minor",
            "assemble2": "patch",
          },
          "summary": "This is a test.",
        },
      ],
      "type": "prerelease",
    },
  },
}
`;

exports[`assemble changes in preMode with no existing changes 1`] = `
Object {
  "changes": Array [
    Object {
      "releases": Object {
        "assemble1": "patch",
        "assemble2": "patch",
      },
      "summary": "This is a test.",
    },
    Object {
      "releases": Object {
        "assemble1": "minor",
        "assemble2": "patch",
      },
      "summary": "This is a test.",
    },
    Object {
      "releases": Object {
        "assemble1": "patch",
        "assemble2": "major",
      },
      "summary": "This is a test.",
    },
    Object {
      "releases": Object {
        "@namespaced/assemble2": "patch",
        "assemble1": "patch",
      },
      "summary": "This is a test. We might link out to a [website](https://www.jacobbolda.com).",
    },
  ],
  "releases": Object {
    "@namespaced/assemble2": Object {
      "changes": Array [
        Object {
          "releases": Object {
            "@namespaced/assemble2": "patch",
            "assemble1": "patch",
          },
          "summary": "This is a test. We might link out to a [website](https://www.jacobbolda.com).",
        },
      ],
      "type": "prepatch",
    },
    "assemble1": Object {
      "changes": Array [
        Object {
          "releases": Object {
            "assemble1": "patch",
            "assemble2": "patch",
          },
          "summary": "This is a test.",
        },
        Object {
          "releases": Object {
            "assemble1": "minor",
            "assemble2": "patch",
          },
          "summary": "This is a test.",
        },
        Object {
          "releases": Object {
            "assemble1": "patch",
            "assemble2": "major",
          },
          "summary": "This is a test.",
        },
        Object {
          "releases": Object {
            "@namespaced/assemble2": "patch",
            "assemble1": "patch",
          },
          "summary": "This is a test. We might link out to a [website](https://www.jacobbolda.com).",
        },
      ],
      "type": "preminor",
    },
    "assemble2": Object {
      "changes": Array [
        Object {
          "releases": Object {
            "assemble1": "patch",
            "assemble2": "patch",
          },
          "summary": "This is a test.",
        },
        Object {
          "releases": Object {
            "assemble1": "minor",
            "assemble2": "patch",
          },
          "summary": "This is a test.",
        },
        Object {
          "releases": Object {
            "assemble1": "patch",
            "assemble2": "major",
          },
          "summary": "This is a test.",
        },
      ],
      "type": "premajor",
    },
  },
}
`;

exports[`assemble changes runs 1`] = `
Object {
  "changes": Array [
    Object {
      "releases": Object {
        "assemble1": "patch",
        "assemble2": "patch",
      },
      "summary": "This is a test.",
    },
    Object {
      "releases": Object {
        "assemble1": "minor",
        "assemble2": "patch",
      },
      "summary": "This is a test.",
    },
    Object {
      "releases": Object {
        "assemble1": "patch",
        "assemble2": "major",
      },
      "summary": "This is a test.",
    },
    Object {
      "releases": Object {
        "@namespaced/assemble2": "patch",
        "assemble1": "patch",
      },
      "summary": "This is a test. We might link out to a [website](https://www.jacobbolda.com).",
    },
  ],
  "releases": Object {
    "@namespaced/assemble2": Object {
      "changes": Array [
        Object {
          "releases": Object {
            "@namespaced/assemble2": "patch",
            "assemble1": "patch",
          },
          "summary": "This is a test. We might link out to a [website](https://www.jacobbolda.com).",
        },
      ],
      "type": "patch",
    },
    "assemble1": Object {
      "changes": Array [
        Object {
          "releases": Object {
            "assemble1": "patch",
            "assemble2": "patch",
          },
          "summary": "This is a test.",
        },
        Object {
          "releases": Object {
            "assemble1": "minor",
            "assemble2": "patch",
          },
          "summary": "This is a test.",
        },
        Object {
          "releases": Object {
            "assemble1": "patch",
            "assemble2": "major",
          },
          "summary": "This is a test.",
        },
        Object {
          "releases": Object {
            "@namespaced/assemble2": "patch",
            "assemble1": "patch",
          },
          "summary": "This is a test. We might link out to a [website](https://www.jacobbolda.com).",
        },
      ],
      "type": "minor",
    },
    "assemble2": Object {
      "changes": Array [
        Object {
          "releases": Object {
            "assemble1": "patch",
            "assemble2": "patch",
          },
          "summary": "This is a test.",
        },
        Object {
          "releases": Object {
            "assemble1": "minor",
            "assemble2": "patch",
          },
          "summary": "This is a test.",
        },
        Object {
          "releases": Object {
            "assemble1": "patch",
            "assemble2": "major",
          },
          "summary": "This is a test.",
        },
      ],
      "type": "major",
    },
  },
}
`;

exports[`merge config test merges nested bumps 1`] = `
Object {
  "commands": Array [
    Object {
      "command": Array [
        "true",
      ],
      "dependencies": Array [
        "assemble1",
        "all",
      ],
      "errorOnVersionRange": null,
      "manager": undefined,
      "path": "./packages/assemble1",
      "pkg": "assemble1",
      "postcommand": null,
      "precommand": null,
      "type": "minor",
    },
    Object {
      "command": Array [
        "true",
      ],
      "dependencies": Array [
        "all",
      ],
      "errorOnVersionRange": null,
      "manager": undefined,
      "path": "./packages/assemble2",
      "pkg": "assemble2",
      "postcommand": null,
      "precommand": null,
      "type": "minor",
    },
    Object {
      "command": Array [
        "true",
      ],
      "dependencies": undefined,
      "errorOnVersionRange": null,
      "manager": undefined,
      "path": undefined,
      "pkg": "all",
      "postcommand": null,
      "precommand": null,
      "type": "minor",
    },
  ],
  "pipeTemplate": Object {
    "all": Object {
      "name": "all",
      "pipe": Object {
        "pkg": Object {
          "command": true,
          "dependencies": undefined,
          "errorOnVersionRange": null,
          "manager": undefined,
          "path": undefined,
          "pkg": "all",
          "postcommand": null,
          "precommand": null,
        },
        "release": Object {
          "changes": Array [
            Object {
              "releases": Object {
                "all": "minor",
              },
              "summary": "This is a test.",
            },
          ],
          "type": "minor",
        },
      },
    },
    "assemble1": Object {
      "name": "assemble1",
      "pipe": Object {
        "pkg": Object {
          "command": true,
          "dependencies": Array [
            "assemble1",
            "all",
          ],
          "errorOnVersionRange": null,
          "manager": undefined,
          "path": "./packages/assemble1",
          "pkg": "assemble1",
          "postcommand": null,
          "precommand": null,
        },
        "release": Object {
          "changes": Array [
            Object {
              "releases": Object {
                "assemble1": "patch",
              },
              "summary": "This is a test.",
            },
          ],
          "type": "minor",
        },
      },
    },
    "assemble2": Object {
      "name": "assemble2",
      "pipe": Object {
        "pkg": Object {
          "command": true,
          "dependencies": Array [
            "all",
          ],
          "errorOnVersionRange": null,
          "manager": undefined,
          "path": "./packages/assemble2",
          "pkg": "assemble2",
          "postcommand": null,
          "precommand": null,
        },
        "release": Object {
          "changes": Array [
            Object {
              "releases": Object {
                "all": "minor",
              },
              "summary": "This is a test.",
            },
          ],
          "type": "minor",
        },
      },
    },
  },
}
`;

exports[`merge config test merges publish 1`] = `
Array [
  Object {
    "command": Array [
      "npm publish",
    ],
    "dependencies": Array [
      "all",
    ],
    "errorOnVersionRange": null,
    "manager": "javascript",
    "path": "./packages/assemble1",
    "pkg": "assemble1",
    "pkgFile": Object {
      "name": "assemble1",
      "pkg": Object {
        "name": "assemble1",
        "version": "0.0.0",
      },
      "version": "0.0.0",
      "versionMajor": 0,
      "versionMinor": 0,
      "versionPatch": 0,
      "versionPrerelease": null,
    },
    "postcommand": null,
    "precommand": null,
  },
  Object {
    "command": Array [
      "cargo publish",
    ],
    "dependencies": Array [
      "assemble1",
      "all",
    ],
    "errorOnVersionRange": null,
    "manager": "cargo",
    "path": "./packages/namespaced-assemble2",
    "pkg": "@namespaced/assemble1",
    "pkgFile": Object {
      "name": "@namespaced/assemble1",
      "pkg": Object {
        "name": "@namedspaced/assemble2",
        "version": "0.0.0",
      },
      "version": "0.0.0",
      "versionMajor": 0,
      "versionMinor": 0,
      "versionPatch": 0,
      "versionPrerelease": null,
    },
    "postcommand": null,
    "precommand": null,
  },
  Object {
    "command": Array [
      "cargo publish",
    ],
    "dependencies": Array [
      "assemble2",
      "all",
    ],
    "errorOnVersionRange": null,
    "manager": "cargo",
    "path": "./packages/namespaced-assemble2",
    "pkg": "@namespaced/assemble2",
    "pkgFile": Object {
      "name": "@namespaced/assemble2",
      "pkg": Object {
        "name": "@namedspaced/assemble2",
        "version": "0.0.0",
      },
      "version": "0.0.0",
      "versionMajor": 0,
      "versionMinor": 0,
      "versionPatch": 0,
      "versionPrerelease": null,
    },
    "postcommand": null,
    "precommand": null,
  },
]
`;

exports[`merge config test merges version 1`] = `
Object {
  "commands": Array [
    Object {
      "command": Array [
        "cargo version patch",
      ],
      "dependencies": Array [
        "assemble2",
        "all",
      ],
      "errorOnVersionRange": null,
      "manager": "cargo",
      "path": "./packages/namespaced-assemble2",
      "pkg": "@namespaced/assemble2",
      "postcommand": null,
      "precommand": null,
      "type": "patch",
    },
    Object {
      "command": Array [
        "lerna version minor --no-git-tag-version --no-push",
      ],
      "dependencies": Array [
        "all",
      ],
      "errorOnVersionRange": null,
      "manager": "javascript",
      "path": "./packages/assemble1",
      "pkg": "assemble1",
      "postcommand": null,
      "precommand": null,
      "type": "minor",
    },
    Object {
      "command": Array [
        "lerna version major",
      ],
      "dependencies": Array [
        "all",
      ],
      "errorOnVersionRange": null,
      "manager": undefined,
      "path": "./packages/assemble2",
      "pkg": "assemble2",
      "postcommand": null,
      "precommand": null,
      "type": "major",
    },
  ],
  "pipeTemplate": Object {
    "@namespaced/assemble2": Object {
      "name": "@namespaced/assemble2",
      "pipe": Object {
        "pkg": Object {
          "command": "cargo version \${ release.type }",
          "dependencies": Array [
            "assemble2",
            "all",
          ],
          "errorOnVersionRange": null,
          "manager": "cargo",
          "path": "./packages/namespaced-assemble2",
          "pkg": "@namespaced/assemble2",
          "postcommand": null,
          "precommand": null,
        },
        "release": Object {
          "changes": Array [
            Object {
              "releases": Object {
                "@namespaced/assemble2": "patch",
                "assemble1": "patch",
              },
              "summary": "This is a test.",
            },
          ],
          "type": "patch",
        },
      },
    },
    "assemble1": Object {
      "name": "assemble1",
      "pipe": Object {
        "pkg": Object {
          "command": "lerna version \${ release.type } --no-git-tag-version --no-push",
          "dependencies": Array [
            "all",
          ],
          "errorOnVersionRange": null,
          "manager": "javascript",
          "path": "./packages/assemble1",
          "pkg": "assemble1",
          "postcommand": null,
          "precommand": null,
        },
        "release": Object {
          "changes": Array [
            Object {
              "releases": Object {
                "assemble1": "patch",
                "assemble2": "patch",
              },
              "summary": "This is a test.",
            },
            Object {
              "releases": Object {
                "assemble1": "minor",
                "assemble2": "patch",
              },
              "summary": "This is a test.",
            },
            Object {
              "releases": Object {
                "assemble1": "patch",
                "assemble2": "major",
              },
              "summary": "This is a test.",
            },
            Object {
              "releases": Object {
                "@namespaced/assemble2": "patch",
                "assemble1": "patch",
              },
              "summary": "This is a test.",
            },
          ],
          "type": "minor",
        },
      },
    },
    "assemble2": Object {
      "name": "assemble2",
      "pipe": Object {
        "pkg": Object {
          "command": "lerna version \${ release.type }",
          "dependencies": Array [
            "all",
          ],
          "errorOnVersionRange": null,
          "manager": undefined,
          "path": "./packages/assemble2",
          "pkg": "assemble2",
          "postcommand": null,
          "precommand": null,
        },
        "release": Object {
          "changes": Array [
            Object {
              "releases": Object {
                "assemble1": "patch",
                "assemble2": "patch",
              },
              "summary": "This is a test.",
            },
            Object {
              "releases": Object {
                "assemble1": "minor",
                "assemble2": "patch",
              },
              "summary": "This is a test.",
            },
            Object {
              "releases": Object {
                "assemble1": "patch",
                "assemble2": "major",
              },
              "summary": "This is a test.",
            },
          ],
          "type": "major",
        },
      },
    },
  },
}
`;

exports[`merge config test merges version without command 1`] = `
Object {
  "commands": Array [],
  "pipeTemplate": Object {},
}
`;

exports[`merge filtered config test merges publish 1`] = `
Array [
  Object {
    "command": Array [
      "npm publish",
    ],
    "dependencies": Array [
      "all",
    ],
    "errorOnVersionRange": null,
    "manager": "javascript",
    "path": "./packages/assemble1",
    "pkg": "assemble1",
    "pkgFile": Object {
      "name": "assemble1",
      "pkg": Object {
        "name": "assemble1",
        "version": "0.0.0",
      },
      "version": "0.0.0",
      "versionMajor": 0,
      "versionMinor": 0,
      "versionPatch": 0,
      "versionPrerelease": null,
    },
    "postcommand": null,
    "precommand": null,
  },
  Object {
    "command": Array [
      "cargo publish",
    ],
    "dependencies": Array [
      "assemble1",
      "all",
    ],
    "errorOnVersionRange": null,
    "manager": "cargo",
    "path": "./packages/namespaced-assemble2",
    "pkg": "@namespaced/assemble1",
    "pkgFile": Object {
      "name": "@namespaced/assemble1",
      "pkg": Object {
        "name": "@namedspaced/assemble2",
        "version": "0.0.0",
      },
      "version": "0.0.0",
      "versionMajor": 0,
      "versionMinor": 0,
      "versionPatch": 0,
      "versionPrerelease": null,
    },
    "postcommand": null,
    "precommand": null,
  },
]
`;

exports[`merge filtered config test merges version 1`] = `
Object {
  "commands": Array [],
  "pipeTemplate": Object {},
}
`;

exports[`special bump types handles an only noop 1`] = `
Object {
  "changes": Array [
    Object {
      "releases": Object {
        "assemble1": "housekeeping",
        "assemble2": "workflows",
      },
      "summary": "This is a test.",
    },
  ],
  "releases": Object {
    "assemble1": Object {
      "changes": Array [
        Object {
          "releases": Object {
            "assemble1": "housekeeping",
            "assemble2": "workflows",
          },
          "summary": "This is a test.",
        },
      ],
      "type": "noop",
    },
    "assemble2": Object {
      "changes": Array [
        Object {
          "releases": Object {
            "assemble1": "housekeeping",
            "assemble2": "workflows",
          },
          "summary": "This is a test.",
        },
      ],
      "type": "noop",
    },
  },
}
`;

exports[`special bump types valid additional bump types 1`] = `
Object {
  "changes": Array [
    Object {
      "releases": Object {
        "assemble1": "patch",
        "assemble2": "patch",
      },
      "summary": "This is a test.",
    },
    Object {
      "releases": Object {
        "assemble1": "minor",
        "assemble2": "patch",
      },
      "summary": "This is a test.",
    },
    Object {
      "releases": Object {
        "assemble1": "patch",
        "assemble2": "major",
      },
      "summary": "This is a test.",
    },
    Object {
      "releases": Object {
        "@namespaced/assemble2": "patch",
        "assemble1": "patch",
      },
      "summary": "This is a test. We might link out to a [website](https://www.jacobbolda.com).",
    },
    Object {
      "releases": Object {
        "assemble1": "housekeeping",
        "assemble2": "workflows",
      },
      "summary": "This is a test.",
    },
  ],
  "releases": Object {
    "@namespaced/assemble2": Object {
      "changes": Array [
        Object {
          "releases": Object {
            "@namespaced/assemble2": "patch",
            "assemble1": "patch",
          },
          "summary": "This is a test. We might link out to a [website](https://www.jacobbolda.com).",
        },
      ],
      "type": "patch",
    },
    "assemble1": Object {
      "changes": Array [
        Object {
          "releases": Object {
            "assemble1": "patch",
            "assemble2": "patch",
          },
          "summary": "This is a test.",
        },
        Object {
          "releases": Object {
            "assemble1": "minor",
            "assemble2": "patch",
          },
          "summary": "This is a test.",
        },
        Object {
          "releases": Object {
            "assemble1": "patch",
            "assemble2": "major",
          },
          "summary": "This is a test.",
        },
        Object {
          "releases": Object {
            "@namespaced/assemble2": "patch",
            "assemble1": "patch",
          },
          "summary": "This is a test. We might link out to a [website](https://www.jacobbolda.com).",
        },
        Object {
          "releases": Object {
            "assemble1": "housekeeping",
            "assemble2": "workflows",
          },
          "summary": "This is a test.",
        },
      ],
      "type": "minor",
    },
    "assemble2": Object {
      "changes": Array [
        Object {
          "releases": Object {
            "assemble1": "patch",
            "assemble2": "patch",
          },
          "summary": "This is a test.",
        },
        Object {
          "releases": Object {
            "assemble1": "minor",
            "assemble2": "patch",
          },
          "summary": "This is a test.",
        },
        Object {
          "releases": Object {
            "assemble1": "patch",
            "assemble2": "major",
          },
          "summary": "This is a test.",
        },
        Object {
          "releases": Object {
            "assemble1": "housekeeping",
            "assemble2": "workflows",
          },
          "summary": "This is a test.",
        },
      ],
      "type": "major",
    },
  },
}
`;
